File: Tools/canada-tax-calculator.py
import json

def get_float_input(prompt: str) -> float:
    """
    Get a float input from the user.
    
    Args:
        prompt: The prompt to display to the user.
    
    Returns:
        A float value entered by the user.
    """
    while True:
        try:
            return float(input(prompt))
        except ValueError:
            print("Please enter a valid number.")

def get_yes_no_input(prompt: str) -> bool:
    """
    Get a yes/no input from the user.
    
    Args:
        prompt: The prompt to display to the user.
    
    Returns:
        A boolean value (True for yes, False for no).
    """
    while True:
        response = input(prompt).strip().lower()
        if response in ['y', 'yes']:
            return True
        elif response in ['n', 'no']:
            return False
        else:
            print("Please enter 'Y' or 'N'.")

def get_shipping_province():
    """
    Get the shipping province code from the user.
    
    Returns:
        A string representing the province code.
    """
    provinces = ['AB', 'BC', 'MB', 'NB', 'NL', 'NS', 'NT', 'NU', 'ON', 'PE', 'QC', 'SK', 'YT']
    while True:
        province = input("Enter the shipping province code (e.g., ON for Ontario): ").upper()
        if province in provinces:
            return province
        print("Invalid province code. Please try again.")

def get_tax_structure(province):
    """
    Get the tax structure for a given province.
    
    Args:
        province: The province code.
    
    Returns:
        A dictionary containing the tax structure.
    """
    tax_structure = {
        'ON': {'type': 'HST', 'rates': ['HST']},
        'BC': {'type': 'GST_PST', 'rates': ['GST', 'PST']},
        'AB': {'type': 'GST', 'rates': ['GST']},
        'MB': {'type': 'GST_PST', 'rates': ['GST', 'PST']},
        'NB': {'type': 'HST', 'rates': ['HST']},
        'NL': {'type': 'HST', 'rates': ['HST']},
        'NS': {'type': 'HST', 'rates': ['HST']},
        'NT': {'type': 'GST', 'rates': ['GST']},
        'NU': {'type': 'GST', 'rates': ['GST']},
        'PE': {'type': 'HST', 'rates': ['HST']},
        'QC': {'type': 'GST_QST', 'rates': ['GST', 'QST']},
        'SK': {'type': 'GST_PST', 'rates': ['GST', 'PST']},
        'YT': {'type': 'GST', 'rates': ['GST']},
    }
    return tax_structure.get(province)

def get_tax_rates(tax_structure):
    """
    Get the tax rates for a given tax structure.
    
    Args:
        tax_structure: The tax structure dictionary.
    
    Returns:
        A dictionary containing the tax rates.
    """
    rates = {}
    if 'HST' in tax_structure['rates']:
        rates['HST'] = get_float_input("Enter HST rate (%): ")
    elif 'GST' in tax_structure['rates']:
        rates['GST'] = get_float_input("Enter GST rate (%): ")
        if tax_structure['type'] in ['GST_PST', 'GST_QST']:
            rate_name = 'PST' if tax_structure['type'] == 'GST_PST' else 'QST'
            rates[rate_name] = get_float_input(f"Enter {rate_name} rate (%): ")
    return rates

def calculate_tax(items, tax_rates, shipping_cost, shipping_taxable, is_sales_before_tax=False, discount_taxable=True):
    """
    Calculate tax based on provided data.
    
    Args:
        items: A list of item dictionaries with 'price' and 'discount' keys.
        tax_rates: A dictionary containing the tax rates.
        shipping_cost: The cost of shipping.
        shipping_taxable: Boolean indicating if shipping is taxable.
        is_sales_before_tax: Boolean indicating if sales tax is applied before discounts.
        discount_taxable: Boolean indicating if discounts are taxable.
    
    Returns:
        A dictionary containing the following keys:
            - total_tax: Total tax amount
            - total_amount: Total amount including tax
            - item_total: Total amount of items
            - discount_total: Total amount of discounts
            - shipping_tax: Tax amount on shipping
            - tax_breakdown: Dictionary containing tax breakdown for each item
    """
    order_total = 0
    item_total = 0
    tax_total = 0
    discount_total = 0
    tax_breakdown = {}

    # Calculate item totals and discount totals
    for i, item in enumerate(items, 1):
        item_price = item['price']
        item_total += item_price
        
        # Track discount total
        if 'discount' in item:
            discount = item['discount']
            discount_total += discount
    
    # Adjust item_total for conditions 4 and 5
    adjusted_item_total = item_total
    if is_sales_before_tax and not discount_taxable:
        adjusted_item_total -= discount_total

    # Calculate taxes based on conditions
    for i, item in enumerate(items, 1):
        item_price = item['price']
        
        # Calculate taxable amount based on global settings
        if 'discount' in item:
            discount = item['discount']
            
            if is_sales_before_tax:
                # Apply discount before calculating tax
                taxable_amount = item_price - discount
            else:
                # When not calculating tax on sales before discount
                taxable_amount = item_price
                
                # Skip tax on discount if applicable
                if not discount_taxable:
                    # Don't include discount in tax calculation
                    taxable_amount = item_price
        else:
            taxable_amount = item_price

        item_tax = 0
        tax_breakdown[f'item{i}'] = {}
        for tax_type, rate in tax_rates.items():
            tax_amount = taxable_amount * rate / 100
            item_tax += tax_amount
            tax_breakdown[f'item{i}'][f"{tax_type}_amount"] = round(tax_amount, 4)

        tax_total += item_tax

    # Handle shipping tax
    shipping_tax = 0
    if shipping_taxable:
        for tax_type, rate in tax_rates.items():
            tax_amount = shipping_cost * rate / 100
            shipping_tax += tax_amount
            tax_breakdown['shipping'] = {f"{tax_type}_amount": round(tax_amount, 4)}
        tax_total += shipping_tax

    # Calculate final amount with special handling for conditions 4 and 5
    if is_sales_before_tax and not discount_taxable:
        # For conditions 4 and 5
        order_total = adjusted_item_total + tax_total + shipping_cost
    else:
        # Standard calculation for other conditions
        order_total = item_total + tax_total + shipping_cost - discount_total

    return {
        "orderTotal": round(order_total, 2),
        "itemTotal": round(adjusted_item_total, 2),  # Use adjusted item total for display
        "taxTotal": round(tax_total, 2),
        "shippingCost": round(shipping_cost, 2),
        "shippingTax": round(shipping_tax, 2),
        "discountTotal": round(discount_total, 2),
        "taxBreakdown": tax_breakdown
    }

File: static/js/pages/canada_tax_calculator.js
/**
 * canada_tax_calculator.js - Canadian Tax Calculator functionality
 * Handles Canadian tax calculation logic, form interactions, and result display
 */
(function() {
    'use strict';

    const CanadaTaxCalculator = {
        // Tax structure by province
        taxStructure: {
            ON: { type: 'HST', rates: ['HST'] },
            BC: { type: 'GST_PST', rates: ['GST', 'PST'] },
            AB: { type: 'GST', rates: ['GST'] },
            MB: { type: 'GST_PST', rates: ['GST', 'PST'] },
            NB: { type: 'HST', rates: ['HST'] },
            NL: { type: 'HST', rates: ['HST'] },
            NS: { type: 'HST', rates: ['HST'] },
            NT: { type: 'GST', rates: ['GST'] },
            NU: { type: 'GST', rates: ['GST'] },
            PE: { type: 'HST', rates: ['HST'] },
            QC: { type: 'GST_QST', rates: ['GST', 'QST'] },
            SK: { type: 'GST_PST', rates: ['GST', 'PST'] },
            YT: { type: 'GST', rates: ['GST'] },
        },
        
        init() {
            this.cacheElements();
            this.bindEvents();
            this.initializeForm();
        },
        
        cacheElements() {
            this.elements = {
                country: document.getElementById('country'),
                province: document.getElementById('province'),
                taxRates: document.getElementById('taxRates'),
                itemsContainer: document.getElementById('itemsContainer'),
                addItemBtn: document.getElementById('addItemBtn'),
                calculateTaxBtn: document.getElementById('calculateTaxBtn'),
                shippingCost: document.getElementById('shippingCost'),
                shippingTaxable: document.getElementById('shippingTaxable'),
                isSalesBeforeTax: document.getElementById('isSalesBeforeTax'),
                discountTaxable: document.getElementById('discountTaxable'),
                container: document.querySelector('.ca-container')
            };
        },
        
        bindEvents() {
            if (this.elements.province) {
                this.elements.province.addEventListener('change', this.updateTaxRates.bind(this));
            }
            
            if (this.elements.country) {
                this.elements.country.addEventListener('change', this.selectTaxCalculator.bind(this));
            }
            
            if (this.elements.addItemBtn) {
                this.elements.addItemBtn.addEventListener('click', this.addItem.bind(this));
            }
            
            if (this.elements.calculateTaxBtn) {
                this.elements.calculateTaxBtn.addEventListener('click', this.calculateTax.bind(this));
            }
        },
        
        initializeForm() {
            this.updateTaxRates();
            this.addItem(false); // First item should not have delete button
        },
        
        updateTaxRates() {
            const province = this.elements.province.value;
            this.elements.taxRates.innerHTML = '';
            
            if (province) {
                const rates = this.taxStructure[province].rates;
                const fragment = document.createDocumentFragment();

                rates.forEach((rate) => {
                    const div = document.createElement('div');
                    div.className = 'form-group ca-tax-rate-field';
                    let defaultValue = '5'; // Default for GST
                    if (rate === 'HST') {
                        defaultValue = '13';
                    } else if (rate === 'PST') {
                        defaultValue = '7';
                    } else if (rate === 'QST') {
                        defaultValue = '9.975';
                    }

                    div.innerHTML = `
                        <label for="${rate.toLowerCase()}Rate">${rate} Rate (%):</label>
                        <input type="number" id="${rate.toLowerCase()}Rate" class="tax-rate-input" step="0.01" min="0" max="100" value="${defaultValue}" required>
                    `;
                    fragment.appendChild(div);
                });
                this.elements.taxRates.appendChild(fragment);
            }
        },
        
        selectTaxCalculator() {
            const country = this.elements.country.value;
            if (country === 'US') {
                window.location.href = this.elements.country.dataset.usCalculatorUrl;
            } else if (country === 'CA') {
                window.location.href = this.elements.country.dataset.caCalculatorUrl;
            }
        },
        
        addItem(includeDeleteButton = true) {
            const itemDiv = document.createElement('div');
            itemDiv.className = 'item';
            
            let deleteButtonHtml = '';
            if (includeDeleteButton) {
                deleteButtonHtml = '<button type="button" class="delete-ca-item-btn">Delete</button>';
            }
            
            itemDiv.innerHTML = `
                <div class="form-group">
                    <label>Item Price ($):</label>
                    <input type="number" class="itemPrice" step="0.01" min="0" value="0.00" required>
                </div>
                <div class="form-group">
                    <label>Discount Amount ($):</label>
                    <input type="number" class="discountAmount" step="0.01" min="0" value="0.00" required>
                </div>
                ${deleteButtonHtml}
            `;
            this.elements.itemsContainer.appendChild(itemDiv);
            
            // Bind delete button event only if button exists
            if (includeDeleteButton) {
                const deleteBtn = itemDiv.querySelector('.delete-ca-item-btn');
                deleteBtn.addEventListener('click', () => this.deleteItem(itemDiv));
            }
        },
        
        /**
         * Delete an item row (must keep at least 1 item)
         * @param {HTMLElement} itemElement - The item element to delete
         */
        deleteItem(itemElement) {
            const itemElements = document.querySelectorAll('.item');
            
            // Cannot delete if only 1 item remains
            if (itemElements.length <= 1) {
                alert('You must have at least one item.');
                return;
            }
            
            itemElement.remove();
        },
        
        calculateTax() {
            const province = this.elements.province.value;
            if (!province) return;

            const taxRates = {};
            
            this.taxStructure[province].rates.forEach((rate) => {
                const input = document.getElementById(rate.toLowerCase() + 'Rate');
                if (input) {
                    taxRates[rate] = parseFloat(input.value);
                }
            });
            
            const items = Array.from(document.querySelectorAll('.item')).map(
                (itemDiv) => ({
                    price: parseFloat(itemDiv.querySelector('.itemPrice').value) || 0,
                    discount: parseFloat(itemDiv.querySelector('.discountAmount').value) || 0
                })
            );
            
            const shippingCost = parseFloat(this.elements.shippingCost.value) || 0;
            const shippingTaxable = this.elements.shippingTaxable.checked;
            const isSalesBeforeTax = this.elements.isSalesBeforeTax.checked;
            const discountTaxable = this.elements.discountTaxable.checked;
            
            const result = this.calculateTaxResult(
                items,
                taxRates,
                shippingCost,
                shippingTaxable,
                isSalesBeforeTax,
                discountTaxable
            );
            
            this.displayResults(result);
        },
        
        calculateTaxResult(items, taxRates, shippingCost, shippingTaxable, isSalesBeforeTax, discountTaxable) {
            let orderTotal = 0;
            let itemTotal = 0;
            let taxTotal = 0;
            let discountTotal = 0;
            let taxBreakdown = {};
            
            items.forEach((item) => {
                itemTotal += item.price;
                if (item.discount > 0) {
                    discountTotal += item.discount;
                }
            });
            
            let adjustedItemTotal = itemTotal;
            if (isSalesBeforeTax && !discountTaxable) {
                adjustedItemTotal = itemTotal - discountTotal;
            }
            
            items.forEach((item, index) => {
                let taxableAmount = item.price;
                
                if (item.discount > 0) {
                    if (isSalesBeforeTax) {
                        taxableAmount = item.price - item.discount;
                    }
                }
                
                let itemTax = 0;
                taxBreakdown[`item${index + 1}`] = {};
                
                for (const [taxType, rate] of Object.entries(taxRates)) {
                    if (isNaN(rate)) continue;
                    const taxAmount = (taxableAmount * rate) / 100;
                    itemTax += taxAmount;
                    taxBreakdown[`item${index + 1}`][`${taxType}_amount`] = Math.round(taxAmount * 10000) / 10000;
                }
                taxTotal += itemTax;
            });
            
            let shippingTax = 0;
            if (shippingTaxable) {
                taxBreakdown['shipping'] = {};
                for (const [taxType, rate] of Object.entries(taxRates)) {
                     if (isNaN(rate)) continue;
                    const taxAmount = (shippingCost * rate) / 100;
                    shippingTax += taxAmount;
                    taxBreakdown['shipping'][`${taxType}_amount`] = Math.round(taxAmount * 10000) / 10000;
                }
                taxTotal += shippingTax;
            }
            
            if (isSalesBeforeTax && !discountTaxable) {
                orderTotal = adjustedItemTotal + taxTotal + shippingCost;
            } else {
                orderTotal = itemTotal + taxTotal + shippingCost - discountTotal;
            }
            
            return {
                orderTotal: Math.round(orderTotal * 100) / 100,
                itemTotal: Math.round(adjustedItemTotal * 100) / 100,
                taxTotal: Math.round(taxTotal * 100) / 100,
                shippingCost: Math.round(shippingCost * 100) / 100,
                shippingTax: Math.round(shippingTax * 100) / 100,
                discountTotal: Math.round(discountTotal * 100) / 100,
                taxBreakdown: taxBreakdown,
            };
        },
        
        displayResults(result) {
            let resultsDiv = document.getElementById('results');
            if (!resultsDiv) {
                resultsDiv = document.createElement('div');
                resultsDiv.id = 'results';
                this.elements.container.appendChild(resultsDiv);
            }
            
            resultsDiv.innerHTML = `
                <h2>Calculation Results:</h2>
                <p>Order Total: $${result.orderTotal.toFixed(2)}</p>
                <p>Item Total: $${result.itemTotal.toFixed(2)}</p>
                <p>Tax Total: $${result.taxTotal.toFixed(2)}</p>
                <p>Shipping Cost: $${result.shippingCost.toFixed(2)}</p>
                <p>Shipping Tax: $${result.shippingTax.toFixed(2)}</p>
                <p>Discount Total: $${result.discountTotal.toFixed(2)}</p>
                <h3>Tax Breakdown:</h3>
                ${Object.entries(result.taxBreakdown)
                    .map(
                        ([itemKey, taxes]) => `
                    <p>${itemKey === 'shipping' ? 'Shipping' : `Item ${itemKey.slice(4)}`} Taxes:</p>
                    <ul>
                        ${Object.entries(taxes)
                            .map(
                                ([taxType, amount]) => `
                            <li>${taxType.toUpperCase().replace('_AMOUNT', '')}: $${amount.toFixed(2)}</li>
                        `
                            )
                            .join('')}
                    </ul>
                `
                    )
                    .join('')}
            `;
        }
    };

    // Module initialization
    document.addEventListener('DOMContentLoaded', function() {
        CanadaTaxCalculator.init();
    });
})();
